x-pack/plugins/apm/server/lib/search_strategies/failure_correlations/async_search_service.ts

```diff
@@ -0,0 +1,129 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import type { ElasticsearchClient } from 'src/core/server';
+import { chunk } from 'lodash';
+import type { SearchServiceParams } from '../../../../common/search_strategies/correlations/types';
+import type { ApmIndicesConfig } from '../../settings/apm_indices/get_apm_indices';
+import { asyncSearchServiceLogProvider } from '../correlations/async_search_service_log';
+import { asyncErrorCorrelationsSearchServiceStateProvider } from './async_search_service_state';
+import { fetchTransactionDurationFieldCandidates } from '../correlations/queries';
+import type { SearchServiceFetchParams } from '../../../../common/search_strategies/correlations/types';
+import { fetchFailureCorrelationPValues } from './queries/query_failure_correlation';
+import { ERROR_CORRELATION_THRESHOLD } from './constants';
+
+export const asyncErrorCorrelationSearchServiceProvider = (
+  esClient: ElasticsearchClient,
+  getApmIndices: () => Promise<ApmIndicesConfig>,
+  searchServiceParams: SearchServiceParams,
+  includeFrozen: boolean
+) => {
+  const { addLogMessage, getLogMessages } = asyncSearchServiceLogProvider();
+
+  const state = asyncErrorCorrelationsSearchServiceStateProvider();
+
+  async function fetchErrorCorrelations() {
+    let params: SearchServiceFetchParams | undefined;
+
+    try {
+      const indices = await getApmIndices();
+      params = {
+        ...searchServiceParams,
+        index: indices['apm_oss.transactionIndices'],
+        includeFrozen,
+      };
+
+      const { fieldCandidates } = await fetchTransactionDurationFieldCandidates(
+        esClient,
+        params
+      );
+
+      addLogMessage(`Identified ${fieldCandidates.length} fieldCandidates.`);
+
+      state.setProgress({ loadedFieldCandidates: 1 });
+
+      let fieldCandidatesFetchedCount = 0;
+      if (params && fieldCandidates.length > 0) {
+        const batches = chunk(fieldCandidates, 10);
+        for (let i = 0; i < batches.length; i++) {
+          try {
+            const results = await Promise.allSettled(
+              batches[i].map((fieldName) =>
+                fetchFailureCorrelationPValues(esClient, params!, fieldName)
```

Comment:
Suggest to try to get rid of the `!`.

Reply:
Updated here https://github.com/elastic/kibana/pull/108441/commits/155443605ee491b2eb7c2bcb7b3e458eed9591e6